[CI] WriteLoadConstraintDeciderIT testHighNodeWriteLoadPreventsNewShardAllocation failing

**Build Scans:**
- [elasticsearch-intake #33914 / part1](https://gradle-enterprise.elastic.co/s/qff7vvjmrhiae)
- [elasticsearch-pull-request #117219 / part-1](https://gradle-enterprise.elastic.co/s/6s2uynlzppw6i)
- [elasticsearch-pull-request #117226 / part-1](https://gradle-enterprise.elastic.co/s/joagd47ffyidc)
- [elasticsearch-pull-request #117219 / part-1](https://gradle-enterprise.elastic.co/s/cxowvnpsc2kxu)

**Reproduction Line:**
```
./gradlew ":server:internalClusterTest" --tests "org.elasticsearch.cluster.routing.allocation.decider.WriteLoadConstraintDeciderIT.testHighNodeWriteLoadPreventsNewShardAllocation" -Dtests.seed=1A79CF871A7CF5E2 -Dtests.locale=pt-PT -Dtests.timezone=US/Eastern -Druntime.java=25
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.cluster.routing.allocation.decider.WriteLoadConstraintDeciderIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testHighNodeWriteLoadPreventsNewShardAllocation'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: safeGet: listener was completed exceptionally
```

**Issue Reasons:**
- [main] 4 failures in test testHighNodeWriteLoadPreventsNewShardAllocation (1.8% fail rate in 227 executions)
- [main] 3 failures in step part-1 (2.2% fail rate in 137 executions)
- [main] 2 failures in pipeline elasticsearch-pull-request (1.5% fail rate in 133 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -333,9 +333,6 @@ tests:
 - class: org.elasticsearch.xpack.security.authc.TokenServiceTests
   method: testAttachAndGetToken
   issue: https://github.com/elastic/elasticsearch/issues/141470
-- class: org.elasticsearch.cluster.routing.allocation.decider.WriteLoadConstraintDeciderIT
-  method: testHighNodeWriteLoadPreventsNewShardAllocation
-  issue: https://github.com/elastic/elasticsearch/issues/141493
 
 # Examples:
 #

--- a/server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderIT.java
@@ -688,36 +688,17 @@ public class WriteLoadConstraintDeciderIT extends ESIntegTestCase {
         int secondDataNodeExpectedNumShards,
         int thirdDataNodeExpectedNumShards
     ) {
-        logger.info("--> Checking shard assignment for index: " + index.getName());
 
         int firstDataNodeRealNumberOfShards = routingNodes.node(firstDataNodeId).numberOfOwningShardsForIndex(index);
         if (firstDataNodeRealNumberOfShards != firstDataNodeExpectedNumShards) {
-            logger.info(
-                "--> First data node has "
-                    + firstDataNodeRealNumberOfShards
-                    + " shards assigned, expected "
-                    + firstDataNodeExpectedNumShards
-            );
             return false;
         }
         int secondDataNodeRealNumberOfShards = routingNodes.node(secondDataNodeId).numberOfOwningShardsForIndex(index);
         if (secondDataNodeRealNumberOfShards != secondDataNodeExpectedNumShards) {
-            logger.info(
-                "--> Second data node has "
-                    + secondDataNodeRealNumberOfShards
-                    + " shards assigned, expected "
-                    + secondDataNodeExpectedNumShards
-            );
             return false;
         }
         int thirdDataNodeRealNumberOfShards = routingNodes.node(thirdDataNodeId).numberOfOwningShardsForIndex(index);
         if (thirdDataNodeRealNumberOfShards != thirdDataNodeExpectedNumShards) {
-            logger.info(
-                "--> Third data node has "
-                    + thirdDataNodeRealNumberOfShards
-                    + " shards assigned, expected "
-                    + thirdDataNodeExpectedNumShards
-            );
             return false;
         }
 

--- a/server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderIT.java
@@ -688,17 +688,36 @@ public class WriteLoadConstraintDeciderIT extends ESIntegTestCase {
         int secondDataNodeExpectedNumShards,
         int thirdDataNodeExpectedNumShards
     ) {
+        logger.info("--> Checking shard assignment for index: " + index.getName());
 
         int firstDataNodeRealNumberOfShards = routingNodes.node(firstDataNodeId).numberOfOwningShardsForIndex(index);
         if (firstDataNodeRealNumberOfShards != firstDataNodeExpectedNumShards) {
+            logger.info(
+                "--> First data node has "
+                    + firstDataNodeRealNumberOfShards
+                    + " shards assigned, expected "
+                    + firstDataNodeExpectedNumShards
+            );
             return false;
         }
         int secondDataNodeRealNumberOfShards = routingNodes.node(secondDataNodeId).numberOfOwningShardsForIndex(index);
         if (secondDataNodeRealNumberOfShards != secondDataNodeExpectedNumShards) {
+            logger.info(
+                "--> Second data node has "
+                    + secondDataNodeRealNumberOfShards
+                    + " shards assigned, expected "
+                    + secondDataNodeExpectedNumShards
+            );
             return false;
         }
         int thirdDataNodeRealNumberOfShards = routingNodes.node(thirdDataNodeId).numberOfOwningShardsForIndex(index);
         if (thirdDataNodeRealNumberOfShards != thirdDataNodeExpectedNumShards) {
+            logger.info(
+                "--> Third data node has "
+                    + thirdDataNodeRealNumberOfShards
+                    + " shards assigned, expected "
+                    + thirdDataNodeExpectedNumShards
+            );
             return false;
         }
 
@@ -779,31 +798,36 @@ public class WriteLoadConstraintDeciderIT extends ESIntegTestCase {
         // Randomly distribute shards' peak write-loads so that we can check later that shard movements are prioritized correctly
         final double writeLoadThreshold = maximumShardWriteLoad
             * BalancedShardsAllocator.Balancer.PrioritiseByShardWriteLoadComparator.THRESHOLD_RATIO;
-        final List<Double> shardPeakWriteLoads = new ArrayList<>();
+        final List<Double> shardRecentWriteLoads = new ArrayList<>();
         // Need at least one with the maximum write-load
-        shardPeakWriteLoads.add((double) maximumShardWriteLoad);
+        shardRecentWriteLoads.add((double) maximumShardWriteLoad);
         final int remainingShards = indexMetadata.getNumberOfShards() - 1;
         // Some over-threshold, some under
         for (int i = 0; i < remainingShards; ++i) {
             if (randomBoolean()) {
-                shardPeakWriteLoads.add(randomDoubleBetween(writeLoadThreshold, maximumShardWriteLoad, true));
+                shardRecentWriteLoads.add(randomDoubleBetween(writeLoadThreshold, maximumShardWriteLoad, true));
             } else {
-                shardPeakWriteLoads.add(randomDoubleBetween(0.0, writeLoadThreshold, true));
+                shardRecentWriteLoads.add(randomDoubleBetween(0.0, writeLoadThreshold, true));
             }
         }
-        assertThat(shardPeakWriteLoads, hasSize(indexMetadata.getNumberOfShards()));
-        Collections.shuffle(shardPeakWriteLoads, random());
+        assertThat(shardRecentWriteLoads, hasSize(indexMetadata.getNumberOfShards()));
+        Collections.shuffle(shardRecentWriteLoads, random());
         final List<ShardStats> shardStats = new ArrayList<>(indexMetadata.getNumberOfShards());
         for (int i = 0; i < indexMetadata.getNumberOfShards(); i++) {
-            shardStats.add(createShardStats(indexMetadata, i, shardPeakWriteLoads.get(i), assignedShardNodeId));
+            shardStats.add(createShardStats(indexMetadata, i, shardRecentWriteLoads.get(i), assignedShardNodeId));
         }
         return shardStats;
     }
 
     /**
-     * Helper to create a dummy {@link ShardStats} for the given index shard with the supplied {@code peakWriteLoad} value.
+     * Helper to create a dummy {@link ShardStats} for the given index shard with the supplied {@code recentWriteLoad} value.
      */
-    private static ShardStats createShardStats(IndexMetadata indexMeta, int shardIndex, double peakWriteLoad, String assignedShardNodeId) {
+    private static ShardStats createShardStats(
+        IndexMetadata indexMeta,
+        int shardIndex,
+        double recentWriteLoad,
+        String assignedShardNodeId
+    ) {
         ShardId shardId = new ShardId(indexMeta.getIndex(), shardIndex);
         Path path = createTempDir().resolve("indices").resolve(indexMeta.getIndexUUID()).resolve(String.valueOf(shardIndex));
         ShardRouting shardRouting = ShardRouting.newUnassigned(
@@ -819,7 +843,7 @@ public class WriteLoadConstraintDeciderIT extends ESIntegTestCase {
         stats.docs = new DocsStats(100, 0, randomByteSizeValue().getBytes());
         stats.store = new StoreStats();
         stats.indexing = new IndexingStats(
-            new IndexingStats.Stats(1, 1, 1, 1, 1, 1, 1, 1, 1, false, 1, 234, 234, 1000, 0.123, peakWriteLoad)
+            new IndexingStats.Stats(1, 1, 1, 1, 1, 1, 1, 1, 1, false, 1, 234, 234, 1000, recentWriteLoad, 0.123)
         );
         return new ShardStats(shardRouting, new ShardPath(false, path, path, shardId), stats, null, null, null, false, 0);
     }

